How many late days are you using for this assignment? None

In [238]:
# Imports and pip installations (if needed)
# %pip install pandas numpy
import pandas as pd
import numpy as np


# Part 1: Load the dataset

In [239]:
# Load the given datasets
categorical_df = pd.read_csv('chronic_kidney_disease_categorical.csv')
numerical_df = pd.read_csv('chronic_kidney_disease_numerical.csv')


# Print the data
print(categorical_df)
print(numerical_df)


     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: The dataset contains medical data of patients from a hospital over the period of two months. Includes a variety of clinical measurements and indicators related to kidney health. It can be used to predict the chronic kidney disease with class labels ckd, notckd. The features in the given datasets are both categorical and numerical. 

The features are: 

1. Age (numerical)
age in years
2.Blood Pressure (numerical)
bp in mm/ Hg
3. Specific Gravity (nominal)
sg - (1.005,1.010,1.015,1.020,1.025)
4. Albumin (nominal)
al - (0,1, 2,3,4,5)
5. Sugar (nominal)
su - (0,1,2,3,4,5)
6. Red Blood Cells (nominal)
rbc - (normal, abnormal)
7. Pus Cell (nominal)
pc - (normal, abnormal)
8. Pus Cell clumps (nominal)
pcc - (present, notpresent)
9. Bacteria (nominal)
ba - (present, notpresent)
10. Blood Glucose Random (numerical)
bgr in mgs/dl
11. Blood Urea (numerical)
bu in mgs/dl
12. Serum Creatinine (numerical)
sc in mgs/dl
13. Sodium(numerical)
sod in mEq/L
14. Potassium (numerical)
pot in mEq/L
15. Hemoglobin (numerical)
hemo in gms
16. Packed Cell Volume (numerical)
17. White Blood Cell Count (numerical)
we in cells/cumm
18. Red Blood Cell Count (numerical)
rc in millions/cmm
19. Hypertension (nominal)
htn - (yes, no)
20. Diabetes Mellitus (nominal)
dm - (yes,no)
21. Coronary Artery Disease (nominal)
cad - (yes, no)
22. Appetite (nominal)
appet - (good, poor)
23. Pedal Edema (nominal)
pe - (yes,no)
24. Anemia (nominal)
ane - (yes, no)
        
The label is : Class (nominal) class - (ckd,notckd). This tells whether or not the target seems to have chronic kidney disease.

The features in the dataset include both Categorical and Numerical features. Here are some examples:

Categorical examples:
1. Red Blood Cells (rbc - (normal,abnormal))
2. Pus Cell (pc - (normal,abnormal))
3. Bacteria (ba  - (present,notpresent))


Numerical Examples:
1. Age (numerical in years)
2. Blood Pressure (bp in mm/Hg)
3. Pottasium (pot in mEq/L)

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [240]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
numerical_duplicates = numerical_df[numerical_df.duplicated()]
print(f"Total number of duplicate rows: {numerical_duplicates.shape[0]}")

# Drop these duplicate rows
numerical_df = numerical_df.drop_duplicates()

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
categorical_duplicates = categorical_df[categorical_df.duplicated()]
print(f"Total number of duplicate rows: {categorical_duplicates.shape[0]}")

# Drop duplicate rows from the categorical dataset
categorical_df = categorical_df.drop_duplicates()


Total number of duplicate rows: 30
Total number of duplicate rows: 30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

In [241]:
#  Testing to see if ids are present in both datasets
# Assuming the unique ID column is named 'id'
unique_ids_cat = set(categorical_df['unique_id'])
unique_ids_num = set(numerical_df['unique_id'])

print("Number of Unique Ids in categorical dataset")
print(len(unique_ids_cat))
print("Number of Unique Ids in numerical dataset")
print(len(unique_ids_num))

missing_in_num = unique_ids_cat - unique_ids_num
print("The exact missing unique ids:")
print(missing_in_num)

# Outer Merge:
outer_combined_df = numerical_df.merge(categorical_df, left_on='unique_id', right_on='unique_id', how="outer")

print("Example of a row that is missing from one of the datasets looks like in the combined table if OUTER JOIN is used:")
print(outer_combined_df.loc[outer_combined_df['unique_id'] == 208034])


# Inner Merge:
inner_combined_df = numerical_df.merge(categorical_df, left_on='unique_id', right_on='unique_id', how="inner")
print("Example of a row that is missing from one of the datasets looks like in the combined table if INNER JOIN is used:")
print(inner_combined_df.loc[inner_combined_df['unique_id'] == 208034])




Number of Unique Ids in categorical dataset
400
Number of Unique Ids in numerical dataset
390
The exact missing unique ids:
{208034, 240430, 488947, 325238, 254775, 292312, 298297, 663387, 948317, 266335}
Example of a row that is missing from one of the datasets looks like in the combined table if OUTER JOIN is used:
    unique_id  age  bp  bgr  bu  sc  sod  pot  hemo  pcv  ...   pc  \
47     208034  NaN NaN  NaN NaN NaN  NaN  NaN   NaN  NaN  ...  NaN   

           pcc          ba  htn   dm  cad appet   pe ane Target  
47  notpresent  notpresent  yes  yes  yes  good  yes  no    ckd  

[1 rows x 25 columns]
Example of a row that is missing from one of the datasets looks like in the combined table if INNER JOIN is used:
Empty DataFrame
Columns: [unique_id, age, bp, bgr, bu, sc, sod, pot, hemo, pcv, wbcc, rbcc, al, su, rbc, pc, pcc, ba, htn, dm, cad, appet, pe, ane, Target]
Index: []

[0 rows x 25 columns]


Answer: 
As you can see in the code above, the categorical dataset has 10 more unique ids than the numerical dataset. Therefore, all unique ids are not present in both datasets. We calculated this by counting the total number of rows in the unique_id columns of both datasets. 

What the rows missing from one of the datasets looks like in the combined dataset depends on how the datasets are merged, i.e. what join operation is used. 
If we use outer join: then the data of the missing unique id is still added to the table but since it has no numerical values (all missing ids are present in cat and NOT in numerical), it has NaN value for all numerical columns. 

If we use inner join: the rows for the missing unique ids are not added to table itself, that's why it returns a empty dataframe if particualrly searched for. 


For the purpose of clean, complete, and ocnsistent data we will proeed with an inner merge.



In [242]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
combined_df = numerical_df.merge(categorical_df, left_on='unique_id', right_on='unique_id', how="inner")

#Print the combined dataset
print(combined_df)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  ...   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  ...   
3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0  ...   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
385     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0  ...   
386     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   
387     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0  ...   
388     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
389     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0  ...   

           pc         pcc          ba  htn   dm cad

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [243]:
# Calculate the percentage of rows that contain atleast one missing value
missing_percentage = combined_df.isnull().any(axis=1).mean() * 100
# Print %
print(missing_percentage)

# Drop these rows from the dataset
cleaned_df = combined_df.dropna()

# Print the Dataset
print(cleaned_df)

60.76923076923077
     unique_id   age     bp    bgr    bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0  19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       938027  43.0   60.0  108.0  25.0  1.0  144.0  5.0  17.8  43.0  ...   
2       421471  37.0   60.0  111.0  35.0  0.8  135.0  4.1  16.2  50.0  ...   
7       109053  64.0   60.0  106.0  27.0  0.7  150.0  3.3  14.4  42.0  ...   
13      349892  59.0   70.0  424.0  55.0  1.7  138.0  4.5  12.6  37.0  ...   
..         ...   ...    ...    ...   ...  ...    ...  ...   ...   ...  ...   
379     486397  20.0   70.0  123.0  44.0  1.0  135.0  3.8  14.6  44.0  ...   
381     462892  36.0   80.0   85.0  16.0  1.1  142.0  4.1  15.6  44.0  ...   
386     475200  64.0  100.0  163.0  54.0  7.2  140.0  4.6   7.9  26.0  ...   
387     168242  48.0   60.0  112.0  44.0  1.2  142.0  4.9  14.5  44.0  ...   
388     474407  21.0   90.0  107.0  40.0  1.7  125.0  3.5   8.3  23.0  ...   

           pc         pcc          ba  htn   

## Part 3.4: Sort the dataset according to the Labels

In [244]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted_df = cleaned_df.sort_values(by=['Target'])

# Reset Index
sorted_df.reset_index(drop=True, inplace=True)

# Print the dataset
print(sorted_df)


     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
2       484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  29.0  ...   
3       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
149     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
150     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
151     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
152     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [245]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']

# get_dummies to sepereate the columns
dummy_df = pd.get_dummies(sorted_df,columns=categorical_columns, prefix=categorical_columns,)

# changing the boolean values to binary
dummy_columns = [col for col in dummy_df.columns if '_' in col and col!="unique_id"]
dummy_df[dummy_columns] = dummy_df[dummy_columns].astype(int)

# Print the dataset
print(dummy_df)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
2       484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  29.0  ...   
3       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
149     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
150     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
151     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
152     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: For binary variables, a single column with 0s and 1s works well for modeling. For categorical variables with more than two categories, one-hot encoding is used, creating separate binary columns for each category. This allows the model to treat each category individually. However, one-hot encoding can increase feature space and sparsity, which might complicate models with many categories. It also treats all categories equally, which can be problematic for ordinal data.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [ ]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

nooutlier_df = dummy_df.copy()

for col in numerical_columns:
    mean = dummy_df[col].mean()
    std_dev = dummy_df[col].std()

    threshold_upper = mean + (3 * std_dev)
    threshold_lower = mean - (3 * std_dev)

    nooutlier_df = nooutlier_df[(nooutlier_df[col] <= threshold_upper) & (nooutlier_df[col] >= threshold_lower)]

# restting index
nooutlier_df.reset_index(drop=True, inplace=True)

# Print the dataset
print(nooutlier_df)


     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
2       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
3       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
4       546225  46.0   60.0  163.0   92.0  3.3  141.0  4.0   9.8  28.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
129     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
130     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
131     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
132     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
133     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [247]:
# Normalize the all Numerical Attributes in the dataset.
df_min_max_scaled = nooutlier_df.copy()

for column in numerical_columns:
    df_min_max_scaled[column] = (df_min_max_scaled[column] - df_min_max_scaled[column].min()) / (df_min_max_scaled[column].max() - df_min_max_scaled[column].min())

# Print the dataset
print(df_min_max_scaled)



     unique_id       age    bp       bgr        bu        sc       sod  \
0       474407  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
1       137148  0.905405  1.00  0.965665  0.522876  0.641975  0.666667   
3       343710  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
4       532520  0.729730  0.75  0.150215  0.281046  0.234568  0.533333   
6       546225  0.540541  0.00  0.399142  0.535948  0.358025  0.700000   
..         ...       ...   ...       ...       ...       ...       ...   
148     647552  0.743243  0.25  0.214592  0.124183  0.037037  0.566667   
149     125587  0.418919  0.00  0.167382  0.241830  0.086420  0.700000   
150     614376  0.689189  0.50  0.210300  0.045752  0.098765  0.500000   
151     853426  0.472973  0.25  0.236052  0.183007  0.024691  0.666667   
152     572140  0.662162  0.50  0.145923  0.117647  0.061728  0.733333   

          pot      hemo       pcv  ...  cad_no  cad_yes  appet_good  \
0    0.171429  0.221311  0.184211  ...  

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:

In [248]:
# Remove that column
df = nooutlier_df.drop(['rbc_abnormal', 'pc_abnormal', 'pcc_notpresent', 'ba_notpresent', 'htn_no', 'dm_no', 'cad_no', 'appet_poor', 'pe_no','ane_no', 'Target_notckd'], axis=1)

# Print the dataset
print(df)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
3       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
6       546225  46.0   60.0  163.0   92.0  3.3  141.0  4.0   9.8  28.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
149     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
150     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
151     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
152     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

     pc_normal  pcc_present  ba_present  htn_yes  d

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [ ]:
# Export the dataframe to a new csv file
df.to_csv('cleaned_data.csv',index=False)

# Export the dataframe to a new json file
df.to_json('cleaned_data.json', orient='columns', index=False)


# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

Paste here: 

| unique_id | age   | bp    | bgr   | bu    | sc   | sod   | pot   | hemo | pcv   | pc_normal | pcc_present | ba_present | htn_yes | dm_yes | cad_yes | appet_good | pe_yes | ane_yes | Target_ckd |
|-----------|-------|-------|-------|-------|------|-------|-------|------|-------|-----------|-------------|------------|---------|--------|---------|------------|--------|---------|------------|
| 474407    | 21.0  | 90.0  | 107.0 | 40.0  | 1.7  | 125.0 | 3.5   | 8.3  | 23.0  | 0         | 1           | 1          | 0       | 0      | 0       | 1          | 0      | 1       | 1          |
| 137148    | 73.0  | 100.0 | 295.0 | 90.0  | 5.6  | 140.0 | 2.9   | 9.2  | 30.0  | 0         | 1           | 0          | 1       | 1      | 1       | 0          | 0      | 0       | 1          |
| 343710    | 61.0  | 80.0  | 173.0 | 148.0 | 3.9  | 135.0 | 5.2   | 7.7  | 24.0  | 0         | 0           | 0          | 1       | 1      | 1       | 0          | 1      | 1       | 1          |
| 532520    | 60.0  | 90.0  | 105.0 | 53.0  | 2.3  | 136.0 | 5.2   | 11.1 | 33.0  | 1         | 0           | 0          | 0       | 0      | 0       | 1          | 0      | 0       | 1          |
| 546225    | 46.0  | 60.0  | 163.0 | 92.0  | 3.3  | 141.0 | 4.0   | 9.8  | 28.0  | 1         | 0           | 0          | 1       | 1      | 0       | 1          | 0      | 0       | 1          |
| 514721    | 56.0  | 90.0  | 129.0 | 107.0 | 6.7  | 131.0 | 4.8   | 9.1  | 29.0  | 0         | 0           | 0          | 1       | 0      | 0       | 1          | 0      | 0       | 1          |
| 621332    | 59.0  | 80.0  | 303.0 | 35.0  | 1.3  | 122.0 | 3.5   | 10.4 | 35.0  | 1         | 0           | 0          | 0       | 1      | 0       | 0          | 0      | 0       | 1          |
| 481293    | 60.0  | 60.0  | 288.0 | 36.0  | 1.7  | 130.0 | 3.0   | 7.9  | 25.0  | 0         | 1           | 0          | 1       | 0      | 0       | 0          | 1      | 0       | 1          |
| 297893    | 68.0  | 80.0  | 157.0 | 90.0  | 4.1  | 130.0 | 6.4   | 5.6  | 16.0  | 0         | 1           | 1          | 1       | 1      | 1       | 0          | 1      | 0       | 1          |
| 615697    | 6.0   | 60.0  | 94.0  | 67.0  | 1.0  | 135.0 | 4.9   | 9.9  | 30.0  | 0         | 0           | 1          | 0       | 0      | 0       | 0          | 0      | 0       | 1          |
| 843362    | 71.0  | 60.0  | 118.0 | 125.0 | 5.3  | 136.0 | 4.9   | 11.4 | 35.0  | 1         | 0           | 0          | 1       | 1      | 0       | 0          | 1      | 0       | 1          |
| 995177    | 69.0  | 70.0  | 214.0 | 96.0  | 6.3  | 120.0 | 3.9   | 9.4  | 28.0  | 0         | 1           | 1          | 1       | 1      | 1       | 1          | 1      | 1       | 1          |
| 828592    | 71.0  | 70.0  | 219.0 | 82.0  | 3.6  | 133.0 | 4.4   | 10.4 | 33.0  | 0         | 1           | 1          | 1       | 1      | 1       | 1          | 0      | 0       | 1          |
| 923613    | 64.0  | 60.0  | 239.0 | 58.0  | 4.3  | 137.0 | 5.4   | 9.5  | 29.0  | 0         | 0           | 1          | 1       | 1      | 0       | 0          | 1      | 0       | 1          |
| 955830    | 55.0  | 80.0  | 214.0 | 73.0  | 3.9  | 137.0 | 4.9   | 10.9 | 34.0  | 0         | 0           | 1          | 1       | 1      | 0       | 1          | 1      | 0       | 1          |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

*We can counter this by cross verifying our data with th eoriginal data and monitor the ocnversions, possibly checking a few manually and see if the outut of the LLM matches. ALso,  implement checks such as count verification. Use specialized tools if possible and make sure hte instructions, inputs, and guidelines are clearly prompted to the LLM model. Prompt engineering is very important in such cases.*

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [250]:
# Code to rename all the columns in the dataset
print("Initial column names")
print(df.columns)

# Column name mapping
abbreviation_mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}

# Function to rename columns
def rename_columns(column_name):
    for abbr, full_name in abbreviation_mapping.items():
        if column_name.startswith(abbr):
            # Replace the abbreviation with the full name, while preserving the suffix
            return column_name.replace(abbr, full_name, 1)  # Replace only the first occurrence
    return column_name

# Rename the columns in the DataFrame
df.columns = [rename_columns(col) for col in df.columns]

# Display the renamed columns
print("Renamed Columns:")
print(df.columns)


Initial column names
Index(['unique_id', 'age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo',
       'pcv', 'wbcc', 'rbcc', 'al', 'su', 'rbc_normal', 'pc_normal',
       'pcc_present', 'ba_present', 'htn_yes', 'dm_yes', 'cad_yes',
       'appet_good', 'pe_yes', 'ane_yes', 'Target_ckd'],
      dtype='object')
Renamed Columns:
Index(['unique_id', 'age', 'blood pressure', 'blood glucose random',
       'blood urea', 'serum creatinine', 'sodium', 'potassium', 'hemoglobin',
       'pus cellv', 'wbcc', 'red blood cellsc', 'albumin', 'sugar',
       'red blood cells_normal', 'pus cell_normal', 'pus cellc_present',
       'bacteria_present', 'hypertension_yes', 'diabetes mellitus_yes',
       'coronary artery disease_yes', 'appetite_good', 'pedal edema_yes',
       'anemia_yes', 'Target_ckd'],
      dtype='object')


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [251]:
# Converted SQL to Pandas code

# Group by the 'Target' column and count the occurrences
result = df.groupby('Target_ckd').size().reset_index(name='count')

# Display the result
print(result)


   Target_ckd  count
0           0    111
1           1     23


*The code above gives us a count of the unique values in the target class grouped by the value of Target_ckd. That is A total count of the number of patients that have chronic kidney disease and another row for the number of people that don't have a chronic kidney disease*. 